### PR TITLE
Filter out abstract plugin classes/interfaces

### PR DESCRIPTION
**Pull request by wendigo**
This change allows plugin class to inherit from abstract plugin class/interface

### CODE REVIEW

File: src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java [hunk 1]
```diff
@@ -127,7 +128,7 @@ private URLClassLoader createClassloaderFromCompileTimeDependencies()
             try {
                 Class<?> implementation = searchRealm.loadClass(pluginClassName);
                 Class<?> clazz = searchRealm.loadClass(className);
-                if (implementation.isAssignableFrom(clazz)) {
+                if (implementation.isAssignableFrom(clazz) && !Modifier.isAbstract(clazz.getModifiers())) {
```
Comment (kokosing) [inline src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java]: static import

File: src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasePlugin.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package its;
+
+import io.prestosql.spi.Plugin;
+
+public abstract class BasePlugin
```
Comment (kokosing) [inline src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasePlugin.java]: BasePlugin -> AbstractPlugin

File: src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasicPlugin.java [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package its;
+
+import io.prestosql.spi.Plugin;
+
+public class BasicPlugin
```
Comment (kokosing) [inline src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasicPlugin.java]: BasicPlugin -> TestPlugin

File: src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasicPlugin.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package its;
```
Comment (kokosing) [inline src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasicPlugin.java]: copyright?

File: src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasePlugin.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package its;
```
Comment (kokosing) [inline src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasePlugin.java]: copyright?

File: src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasicPlugin.java (hunk 4 above)
Comment (wendigo) [reply src/test/projects/basic-plugin-hierarchy/src/main/java/its/BasicPlugin.java]: this are test cases classes (they don't have a license)

File: src/test/java/io/prestosql/maven/CheckerIntegrationTest.java [hunk 6]
```diff
@@ -37,6 +37,16 @@ public void testBasic()
                 .assertErrorFreeLog();
     }
 
+    @Test
+    public void testPluginHierarchy()
```
Comment (findepi) [inline src/test/java/io/prestosql/maven/CheckerIntegrationTest.java]: i would consider renaming to be more explicit:

testPluginHierarchy  -> `testAbstractPluginClass` 
basic-plugin-hierarchy -> abstract-plugin-class

Comment (wendigo) [reply src/test/java/io/prestosql/maven/CheckerIntegrationTest.java]: ok

**Comment by wendigo** [comment]
cc @electrum

File: src/test/java/io/prestosql/maven/GeneratorIntegrationTest.java [hunk 7]
```diff
@@ -58,26 +72,26 @@ public void testBasic()
...
+        File servicesJarFile = new File(basedir, format("target/%s-1.0-services.jar", projectId));
         assertThat(servicesJarFile).isFile();
 
         try (JarFile jar = new JarFile(servicesJarFile)) {
             JarEntry entry = jar.getJarEntry(DESCRIPTOR);
             assertNotNull(entry);
             try (InputStream in = jar.getInputStream(entry)) {
                 String contents = new String(toByteArray(in), UTF_8);
-                assertThat(contents).isEqualTo("its.BasicPlugin\n");
+                assertThat(contents).isEqualTo(format("%s\n", expectedPluginClass));
```
Comment (electrum) [inline src/test/java/io/prestosql/maven/GeneratorIntegrationTest.java]: We usually avoid `format` for a single concat operation
```java
assertThat(contents).isEqualTo(expectedPluginClass + "\n");

File: src/test/java/io/prestosql/maven/GeneratorIntegrationTest.java [hunk 8]
```diff
@@ -58,26 +72,26 @@ public void testBasic()
...
-        File pluginZipFile = new File(basedir, "target/basic-1.0.zip");
+        File pluginZipFile = new File(basedir, format("target/%s-1.0.zip", projectId));
         assertThat(pluginZipFile).isFile();
 
         try (ZipFile zip = new ZipFile(pluginZipFile)) {
             assertThat(list(zip.entries()))
                     .extracting(ZipEntry::getName)
-                    .contains("basic-1.0/basic-1.0.jar")
-                    .contains("basic-1.0/basic-1.0-services.jar");
+                    .contains(format("%s-1.0/%s-1.0.jar", projectId, projectId))
```
Comment (electrum) [inline src/test/java/io/prestosql/maven/GeneratorIntegrationTest.java]: It seems cleaner to reference the same input twice, to show that it's supposed to be the same.
```java
format("%1$s-1.0/%1$s-1.0.jar", projectId)

File: src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java [hunk 9]
```diff
@@ -127,7 +128,7 @@ private URLClassLoader createClassloaderFromCompileTimeDependencies()
             try {
                 Class<?> implementation = searchRealm.loadClass(pluginClassName);
                 Class<?> clazz = searchRealm.loadClass(className);
-                if (implementation.isAssignableFrom(clazz)) {
+                if (implementation.isAssignableFrom(clazz) && !isAbstract(clazz.getModifiers())) {
```
Comment (electrum) [inline src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java]: Does this also exclude interfaces? e.g.
```java
public interface MyPlugin extends Plugin {}

Comment (wendigo) [reply src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java]: No, there was no case for that. I can add it too if you want me to :)

Comment (wendigo) [reply src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java]: I've added this case with tests

File: src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java [hunk 10]
```diff
@@ -148,4 +149,9 @@ private static void mkdirs(File file)
             throw new MojoExecutionException(format("%n%nFailed to create directory: %s", file));
         }
     }
+
+    private static boolean isInstantiable(Class<?> pluginClass, Class<?> clazz)
```
Comment (findepi) [inline src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java]: I don't like the name of the method. `isInstantiable` should not check `pluginClass.isAssignableFrom(clazz)` ...

call it `isImplementation`

also, rename `implementation` variable to `pluginClass`  in the caller scope too

Comment (wendigo) [reply src/main/java/io/prestosql/maven/ServiceDescriptorGenerator.java]: Done

**Comment by findepi** [comment]
@wendigo lgtm. can you please squash the commits?

**Comment by wendigo** [comment]
Squashed @findepi

**Comment by findepi** [comment]
Merged, thanks!

**Pull request merged**